### PR TITLE
Improved API for thermal eigenstrains 

### PR DESCRIPTION
**Pull request by dschwen**
Another breaking change shortly after the last one got in. Sorry about that, but I think I got a way better idea. The flaw of the past and current designs IMO is the returning of values using reference and pointer arguments. We need to shoehorn the AD and non-AD use cases into one function signature, which leads to unused arguments for the derivative in the AD case. This is not elegant.

The new design uses return types to return values only. The return type `ValueAndDerivative<is_ad>` is templated and resolves to `ADReal` for `is_ad == true` and `ChainedReal` for `is_ad == false`. The chained real is a storage for two real values (the thermal expansion and it's temperature derivative). The coolest use of chained real can be seen in [`ComputeDilatationThermalExpansionEigenstrainBase.C`](https://github.com/idaholab/moose/pull/20982/files#diff-d3314902598a119ec1d66662ecff5cc0151ca4e6d8c5f5710e74ccc7ae5cc24b) where through `auto` and the overloaded `-` operator very generic code is possible.

Additionally this PR adds `ChainedReal` support for `LinearInterpolation`s and MOOSE `Function`s.

Refs #20943

### CODE REVIEW

**Comment by dschwen** [comment]
> The coolest use of chained real can be seen in [`ComputeDilatationThermalExpansionEigenstrainBase.C`](https://github.com/idaholab/moose/pull/20982/files#diff-d3314902598a119ec1d66662ecff5cc0151ca4e6d8c5f5710e74ccc7ae5cc24b) where through `auto` and the overloaded `-` operator very generic code is possible.

... and it is wrong (because `stress_free` expression does not depend on temperature and still gets a derivative hat cancels out the `current` term derivative (that actually depends on temperature). Well... work in progress.

File: modules/tensor_mechanics/src/materials/ComputeInstantaneousThermalExpansionFunctionEigenstrain.C [hunk 1]
```diff
@@ -53,36 +53,32 @@ ComputeInstantaneousThermalExpansionFunctionEigenstrainTempl<is_ad>::initQpState
...
-  {
-    mooseAssert(dthermal_strain_dT, "Internal error. dthermal_strain_dT should not be nullptr.");
-    *dthermal_strain_dT = alpha_current_temp;
-  }
+  if constexpr (is_ad)
+    return thermal_strain;
   else
-    libmesh_ignore(dthermal_strain_dT);
+    // return {thermal_strain, 0.5 * (alpha_current_temp + alpha_old_temp)};
+    return {thermal_strain, alpha_current_temp};
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeInstantaneousThermalExpansionFunctionEigenstrain.C:81]: @bwspenc , this derivative seem wrong to me. The derivative should be what it is in the commented out version, but that leads to two tests diffing.

File: modules/tensor_mechanics/include/materials/ComputeDilatationThermalExpansionFunctionEigenstrain.h [hunk 2]
```diff
@@ -27,18 +27,10 @@ class ComputeDilatationThermalExpansionFunctionEigenstrainTempl
 protected:
   /*
    * Compute the fractional linear dilatation due to thermal expansion delta L / L
```
Comment (recuero) [inline modules/tensor_mechanics/include/materials/ComputeDilatationThermalExpansionFunctionEigenstrain.h:29]: Doxygen comment should reflect the templated nature of this method (i.e. does different things for different cases)

File: modules/tensor_mechanics/include/materials/ComputeMeanThermalExpansionEigenstrainBase.h [hunk 3]
```diff
@@ -54,16 +50,8 @@ class ComputeMeanThermalExpansionEigenstrainBaseTempl
    * \f$\bar{\alpha}=(\delta L / L)/(T - T_{ref})\f$.
    * @param temperature  temperature at which this is evaluated
```
Comment (recuero) [inline modules/tensor_mechanics/include/materials/ComputeMeanThermalExpansionEigenstrainBase.h:51]: Some of these doxygen comments appear to need to be updated.

File: modules/tensor_mechanics/src/materials/ComputeDilatationThermalExpansionFunctionEigenstrain.C [hunk 4]
```diff
@@ -36,19 +36,16 @@ ComputeDilatationThermalExpansionFunctionEigenstrainTempl<is_ad>::
 }
 
 template <bool is_ad>
-GenericReal<is_ad>
+ValueAndDerivative<is_ad>
```
Comment (recuero) [inline modules/tensor_mechanics/src/materials/ComputeDilatationThermalExpansionFunctionEigenstrain.C:39]: Are we potentially making copies of `ADReal`s by having the return types without references?

File: modules/tensor_mechanics/include/materials/ComputeThermalExpansionEigenstrainBase.h [hunk 5]
```diff
@@ -28,27 +33,27 @@ class ComputeThermalExpansionEigenstrainBaseTempl
...
 
   /// lag temperature variable
   const bool _use_old_temperature;
 
+  /// temperature to use in the eigenstrain calculation (depending on _use_old_temperature)
+  ValueAndDerivative<is_ad> _temperature;
+
   /// current temperature
-  const GenericVariableValue<is_ad> & _temperature;
+  const GenericVariableValue<is_ad> & _temperature_prop;
```
Comment (recuero) [inline modules/tensor_mechanics/include/materials/ComputeThermalExpansionEigenstrainBase.h:53]: Why `prop`?

Comment (recuero) [reply modules/tensor_mechanics/include/materials/ComputeThermalExpansionEigenstrainBase.h:53]: oh okay

File: modules/tensor_mechanics/src/materials/ComputeDilatationThermalExpansionEigenstrainBase.C [hunk 6]
```diff
@@ -24,31 +24,15 @@ ComputeDilatationThermalExpansionEigenstrainBaseTempl<is_ad>::
...
-
-template <bool is_ad>
-Real
-ComputeDilatationThermalExpansionEigenstrainBaseTempl<is_ad>::computeDilatationDerivative(
-    const Real)
-{
-  mooseError("computeDilatationDerivative must be implemented for any derived non-AD class.");
+  // in non-AD mode the T derivative of the stress_free term needs get dropped.
+  // We assume _stress_free_temperature does not depend on T. In AD mode this is automatic.
+  return current - (is_ad ? stress_free : stress_free.value());
```
Comment (recuero) [inline modules/tensor_mechanics/src/materials/ComputeDilatationThermalExpansionEigenstrainBase.C:35]: I take it this is doing the right thing with `ChainedReal`. It's not immediately apparent. (Just a comment)

**Review by recuero — COMMENTED** [review]
Tests are still making reference to deleted AD docs. I tend to miss this stuff...

File: modules/tensor_mechanics/src/materials/ComputeDilatationThermalExpansionFunctionEigenstrain.C (hunk 4 above)
Comment (dschwen) [reply modules/tensor_mechanics/src/materials/ComputeDilatationThermalExpansionFunctionEigenstrain.C:39]: RVO or NRVO should prevent the copying, and C++17 guarantees copy elision.

**Comment by dschwen** [comment]
> Tests are still making reference to deleted AD docs. I tend to miss this stuff...

Yeah, me, too. I'll do a pass looking for that and verifying that all AD docs were redundant.

File: modules/tensor_mechanics/include/materials/ComputeInstantaneousThermalExpansionFunctionEigenstrain.h [hunk 7]
```diff
@@ -26,14 +26,15 @@ class ComputeInstantaneousThermalExpansionFunctionEigenstrainTempl
...
+  virtual ValueAndDerivative<is_ad> computeThermalStrain() override;
 
   const Function & _thermal_expansion_function;
 
-  /// Stores the thermal strain as a scalar for use in computing an incremental update to this.
-  /// this is not a dual number as only the previous timestep values are used
-  //@{
+  /**
+   *@{ Stores the thermal strain as a scalar for use in computing an incremental update to this.
+   * this is not a dual number as only the previous timestep values are used
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/ComputeInstantaneousThermalExpansionFunctionEigenstrain.h:36]: After talking to @recuero we agree that there is no drawback in making this `GenericReal<is_ad>` (which would make this property potentially more useful and align with user's expectations and ensure interoperability with other objects that might calculate or consume a thermal strain property)

File: modules/tensor_mechanics/include/materials/ComputeThermalExpansionEigenstrainBase.h [hunk 8]
```diff
@@ -26,38 +31,55 @@ class ComputeThermalExpansionEigenstrainBaseTempl
...
+   * a variant of forward mode automatic differentiation will be used internally to
+   * compute the thermal strain and no manual implementation of the temperature derivative
+   * is needed.
    */
-  virtual void computeThermalStrain(GenericReal<is_ad> & thermal_strain,
-                                    Real * dthermal_strain_dT = nullptr) = 0;
+  virtual ValueAndDerivative<is_ad> computeThermalStrain() = 0;
+
+  /**
+   * Temperature to use in the eigenstrain calculation (depending on _use_old_temperature).
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/ComputeThermalExpansionEigenstrainBase.h:56]: I'd make this more explicitly clear:
Temperature used in the eigenstrain calculation (current value if _use_old_temperature=false, old value if _use_old_temperature=true).

**Review by bwspenc — CHANGES_REQUESTED** [review]
Looks good. Just some really minor comments on your comments.

**Review by bwspenc — APPROVED** [review]

**Comment by bwspenc** [comment]
This change set breaks some apps, but PRs are ready on those apps.

**Comment by loganharbour** [comment]
Please notify slack in the future for breaking changes

**Pull request merged**